### PR TITLE
perf: Allow opting out of arg/env expansion

### PR DESCRIPTION
**Pull request by DavidZbarsky-at**
This saves around 2s of analysis time in our repo; we have a few thousand binaries and growing.

### Test plan

- Covered by existing test cases

### CODE REVIEW

File: js/private/js_binary.bzl [hunk 1]
```diff
@@ -322,18 +323,24 @@ _NODE_OPTION = """JS_BINARY__NODE_OPTIONS+=(\"{value}\")"""
...
     if is_windows:
         fixed_env = dict(fixed_env, **{"JS_BINARY__PATCH_NODE_FS": "0"})
 
-    envs = []
-    for (key, value) in dicts.add(fixed_env, ctx.attr.env).items():
-        envs.append(_ENV_SET.format(
-            var = key,
-            value = " ".join([expand_variables(ctx, exp, attribute_name = "env") for exp in expand_locations(ctx, value, ctx.attr.data).split(" ")]),
-        ))
+    envs = [
```
Comment (DavidZbarsky-at) [inline js/private/js_binary.bzl:337]: refactoring this was around 100ms win, but I was already here so might as well

Comment (jbedard) [reply js/private/js_binary.bzl:337]: Worth it's own PR so we can merge now?

Comment (DavidZbarsky-at) [reply js/private/js_binary.bzl:337]: It's a bit wordy without the extracted helper. I'd rather just land this together, don't want to create more work for myself rebasing

File: js/private/js_binary.bzl [hunk 2]
```diff
@@ -126,6 +125,8 @@ _ATTRS = {
         and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution.
         """,
     ),
+    "expand_args": attr.bool(default = True),
```
[the remark this answers is not in the document]
Comment (DavidZbarsky-at) [reply js/private/js_binary.bzl]: Jason had suggested leaving these undocumented but easy enough to add.

I added a test for args not being expanded, but I think adding a test for `env` is a bit more involved - I'm not sure if `js_run_binary` will work without the expansion. I also didn't see existing tests for env expansion, maybe I missed them somewhere?

File: js/private/js_binary.bzl [hunk 3]
```diff
@@ -123,15 +122,29 @@ _ATTRS = {
...
-        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution.
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution if `expand_env` is set to True.
         """,
     ),
+    "expand_args": attr.bool(
+        default = True,
+        doc = """Enables [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution for `fixed_args`.
+
+        This comes at some analysis-time cost.""",
```
Comment (gregmagolan) [inline js/private/js_binary.bzl:133]: nit: This comes at some analysis-time cost even for a set of args that does not have any expansions.

File: js/private/js_binary.bzl [hunk 4]
```diff
@@ -123,15 +122,29 @@ _ATTRS = {
...
+    ),
+    "expand_env": attr.bool(
+        default = True,
+        doc = """Enables [$(location)](https://bazel.build/reference/be/make-variables#predefined_label_variables)
+        and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution for `env`.
+
+        This comes at some analysis-time cost.""",
```
Comment (gregmagolan) [inline js/private/js_binary.bzl:140]: nit: This comes at some analysis-time cost even for a set of envs that does not have any expansions.

**Review by gregmagolan — APPROVED** [review]
🦖

**Review by jbedard — APPROVED** [review]
